File: framework/src/java/wandledi/java/html/Selectable.java
package wandledi.java.html;

import wandledi.core.Attribute;
import wandledi.core.Selector;

public interface Selectable {

    Element get(String selector);

    Element get(String label, String attrName, String attrValue);

    Element get(String attrName, String attrValue);

    Element get(String label, Attribute... attributes);

    Element get(Attribute... attributes);

    /**Retrieves an Element object for the elements matching the given selector.
     *
     * @param selector
     * @return
     */
    Element get(Selector selector);

    Selectable at(Selector selector);

    Selectable at(String selector);
}

File: scala-lib/src/main/scala/wandledi/scala/html/Selectable.scala
package wandledi.scala.html

import wandledi.core.Attribute
import wandledi.core.CssSelector
import wandledi.core.Scroll
import wandledi.core.Selector
import wandledi.core.UniversalSelector
import wandledi.java.html.LocalSpells

class Selectable(scroll: Scroll) extends wandledi.java.html.SelectableImpl(scroll) {

  override def get(selector: Selector) = new Element(selector, scroll)

  def get(atts: Tuple2[String, String]*): Element = {
    val jatts = for (attr <- atts) yield new Attribute(attr._1, attr._2)
    get(new UniversalSelector(jatts: _*))
  }

  def get(label: String, atts: Tuple2[String, String]*): Element = {
    val jatts = for (attr <- atts) yield new Attribute(attr._1, attr._2)
    get(new UniversalSelector(label, jatts: _*))
  }

  override def get(selector: String): Element = get(CssSelector.valueOf(selector))

  override def at(selector: Selector): Selectable = {
    val nestedScroll = new Scroll
    val localSpell = new LocalSpells(scroll, nestedScroll)
    scroll.addSpell(selector, localSpell)
    new Selectable(nestedScroll)
  }

  override def at(selector: String): Selectable = at(CssSelector.valueOf(selector))
}
